src/bp/nlu-core/entities/microsoft-extractor/index.ts

```diff
@@ -0,0 +1,189 @@
+import { Recognizer, IModel, ModelResult } from '@microsoft/recognizers-text'
+
+import { NLU } from 'botpress/sdk'
+import _ from 'lodash'
+
+import { EntityExtractionResult, SystemEntityExtractor, KeyedItem } from '../../typings'
+import { SystemEntityCacheManager } from '../entity-cache-manager'
+import {
+  SupportedLangs,
+  SupportedLangsList,
+  GlobalRecognizers,
+  LanguageDependantRecognizers,
+  DucklingUnitMapping,
+  DucklingDateMappings,
+  DucklingTypeMappings,
+  langToCulture,
+  MicrosoftValues,
+  MicrosoftTimeValues,
+  MicrosoftValue
+} from './typings'
+interface MicrosoftParams {
+  lang: SupportedLangs
+  recognizers: any[]
+}
+
+const BATCH_SIZE = 10000
+
+// Further improvements:
+// 1- in _extractBatch, shift results ==> don't walk whole array n times (nlog(n) vs n2)
+
+export class MicrosoftEntityExtractor implements SystemEntityExtractor {
+  constructor(private _cache: SystemEntityCacheManager, private readonly logger?: NLU.Logger) {
+    this.logger = logger
+  }
+
+  public async configure() {
+    await this._cache.restoreCache()
+  }
+
+  public resetCache() {
+    this._cache.reset()
+  }
+
+  public async extractMultiple(
+    inputs: string[],
+    lang: string,
+    useCache?: boolean
+  ): Promise<EntityExtractionResult[][]> {
+    let options: MicrosoftParams
+
+    if (!SupportedLangsList.includes(lang)) {
+      lang = 'en'
+      options = {
+        lang: 'en' as SupportedLangs,
+        recognizers: [...GlobalRecognizers]
+      }
+    } else {
+      options = {
+        lang: lang as SupportedLangs,
+        recognizers: [...LanguageDependantRecognizers, ...GlobalRecognizers]
+      }
+    }
+
+    const [cached, toFetch] = this._cache.splitCacheHitFromCacheMiss(inputs, !!useCache)
+
+    const chunks = _.chunk(toFetch, BATCH_SIZE)
```

Comment:
just for fun (I just think about this) why do we batch predictions ? Is this recommanded by the lib? We only batched requests in duckling because its reached by HTTP and we can't send too big of a payload.

This one is called in-memory, so why making batches?

Reply:
You're right, stupid copy paste from the duckling code. I removed it and just pass toFetch to the extract fonction.